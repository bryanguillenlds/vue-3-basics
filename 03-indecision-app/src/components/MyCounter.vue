<template>
	<section>
		<h3>{{ count }}</h3>
		<h3>{{ squaredCount }}</h3>

		<div>
			<button @click="count++">+1</button>
			<button @click="count--">-1</button>
		</div>
	</section>
</template>

<script setup lang="ts">
//setup keyword is so that you don't have to use setup in the script tag
//this already tells it what to do

import { ref, computed, defineProps } from 'vue';

/* Option 1: Runtime Props (JavaScript-style with TypeScript)
// PROS:
// - Runtime validation: Vue checks prop types at runtime and throws clear errors
// - Better debugging: You get helpful error messages if wrong props are passed
// - Works without TypeScript: Can be used in regular JavaScript files
// - Built-in validation: Vue handles type checking automatically
// - Clear error messages: "Invalid prop: type check failed for prop 'value'"
//
// CONS:
// - Less type safety: TypeScript can't infer exact types as well
// - Runtime overhead: Slight performance cost due to validation
// - Verbose syntax: More code to write for complex prop definitions
// - Limited TypeScript features: Can't use advanced types like unions easily */
// const props = defineProps({
// 	value: {
// 		type: Number,
// 		required: true
// 	}
// });

/* Option 2: Type-Only Props (Pure TypeScript Generic)
// PROS:
// - Full TypeScript support: Complete type safety and IntelliSense
// - Better performance: No runtime validation overhead
// - Cleaner syntax: Less verbose, more readable
// - Advanced types: Can use unions, generics, complex interfaces
// - Compile-time checking: Catches errors before runtime
// - Better refactoring: TypeScript can track prop usage across files
//
// CONS:
// - No runtime validation: Props aren't checked at runtime
// - TypeScript required: Won't work in JavaScript files
// - Silent failures: Wrong prop types might not cause obvious errors
// - Learning curve: Requires understanding TypeScript generics */
interface Props {
	value: number;
}

const props = defineProps<Props>();

const count = ref(props.value);

const squaredCount = computed(() => {
	return count.value * count.value;
});
</script>

<style scoped>
h1 {
	color: red;
}
</style>
