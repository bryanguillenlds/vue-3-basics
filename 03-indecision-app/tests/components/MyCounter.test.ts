import { mount } from '@vue/test-utils';
import MyCounter from '@/components/MyCounter.vue';

// Testing snapshots are necessary when you want to compare
// the output of the component with a previous output in case something in the HTML changes
describe('MyCounter', () => {
	test('should match snapshot', () => {
		// Mount is used to simulate the component in the DOM
		const wrapper = mount(MyCounter, {
			props: {
				value: 5
			}
		});

		expect(wrapper.html()).toMatchSnapshot();
	});

	test('renders the counter value correctly', () => {
		const wrapper = mount(MyCounter, {
			props: {
				value: 5
			}
		});

		const [count, squaredCount] = wrapper.findAll('h3');

		expect(count.text()).toBe('5');
		expect(squaredCount.text()).toBe('25'); // 5 * 5 = 25
	});

	test('should increment the counter when the increment button is clicked', async () => {
		const wrapper = mount(MyCounter, {
			props: {
				value: 5
			}
		});

		const [count, squaredCount] = wrapper.findAll('h3');

		const btnIncrement = wrapper.find('button');

		await btnIncrement.trigger('click');
		expect(count.text()).toBe('6');
		expect(squaredCount.text()).toBe('36'); // 6 * 6 = 36
	});

	test('should decrement the counter when the decrement button is clicked', async () => {
		const wrapper = mount(MyCounter, {
			props: {
				value: 5
			}
		});

		const [count, squaredCount] = wrapper.findAll('h3');

		const btnDecrement = wrapper.find('button:last-child');

		await btnDecrement.trigger('click');
		expect(count.text()).toBe('4');
		expect(squaredCount.text()).toBe('16'); // 4 * 4 = 16
	});
});
