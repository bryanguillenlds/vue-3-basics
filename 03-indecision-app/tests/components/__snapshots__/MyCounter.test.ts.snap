// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MyCounter > should match snapshot 1`] = `
"<section data-v-0352012b="">
  <h3 data-v-0352012b="">5</h3>
  <h3 data-v-0352012b="">25</h3>
  <div data-v-0352012b="" class="flex gap-2"><button data-v-0352012b="" class="bg-blue-500 text-white p-5 rounded-md mr-2">+1</button><button data-v-0352012b="" class="bg-red-500 text-white p-5 rounded-md">-1</button></div>
</section>"
`;
